### PR TITLE
Added retry to raw+tcp adapter

### PR DESCRIPTION
**Pull request by alex3305**
Most of this change is duplicated from the syslog adapter, since it contains a working retry. Since the raw adapter also doesn't have any tests associated with it, this is not implemented into this version.

Since the documentation already contains the necessary retry information for TCP adapters, I didn't see any need to alter this.

### CODE REVIEW

File: VERSION [hunk 1]
```diff
@@ -1 +1 @@
-v3.2.5
+v3.2.6
```
Comment (michaelshobbs) [inline VERSION]: this is handled at release time. please revert

File: adapters/raw/raw.go [hunk 2]
```diff
@@ -7,14 +7,43 @@ import (
 	"log"
 	"net"
 	"os"
-	"reflect"
-	"text/template"
+		"text/template"
```
Comment (michaelshobbs) [inline adapters/raw/raw.go]: spacing is off. make sure you run your go code through `gofmt`

File: adapters/raw/raw.go [hunk 3]
```diff
@@ -73,9 +104,89 @@ func (a *Adapter) Stream(logstream chan *router.Message) {
...
+		}
+
+		try++
+		if try > tries {
+			return err
+		}
+
+		time.Sleep((1 << try) * 10 * time.Millisecond)
+	}
+}
```
Comment (michaelshobbs) [inline adapters/raw/raw.go]: add newline to end of file please

Comment (alex3305) [reply adapters/raw/raw.go]: Fixed

**Comment by alex3305** [comment]
@michaelshobbs I've amended all your concerns and rebased those in the existing commit. Regarding tests, I'll look into it.

**Comment by alex3305** [comment]
I've amended my commit. Is it possible to re-review this?

**Comment by gbolo** [comment]
Hi @alex3305  please rebase to resolve the conflicts and also run `goimports` on `adapters/raw/raw.go`

**Comment by gbolo** [comment]
Hi @alex3305 
This code does not seem to work as advertised. Needs fixing:

```
2018/11/02 01:09:25 raw: write tcp 127.0.0.1:49294->127.0.0.1:9999: write: broken pipe
2018/11/02 01:09:25 raw: reconnecting up to 10 times
2018/11/02 01:09:25 raw: reconnect successful
2018/11/02 01:09:33 raw: write tcp 127.0.0.1:49310->127.0.0.1:9999: write: broken pipe
2018/11/02 01:09:33 raw: reconnecting up to 10 times
2018/11/02 01:09:54 raw retry err: dial tcp 127.0.0.1:9999: connect: connection refused
panic: raw retry err: dial tcp 127.0.0.1:9999: connect: connection refused

goroutine 54 [running]:
log.Panicf(0x5637c15aed9c, 0x12, 0xc420151f20, 0x1, 0x1)
        /usr/lib/go/src/log/log.go:333 +0xdc
github.com/gliderlabs/logspout/adapters/raw.(*Adapter).Stream(0xc4201d9020, 0xc4203e62a0)
        /go/src/github.com/gliderlabs/logspout/adapters/raw/raw.go:112 +0x2ad
github.com/gliderlabs/logspout/router.(*RouteManager).route(0xc420024440, 0xc4200cbe40)
        /go/src/github.com/gliderlabs/logspout/router/routes.go:160 +0xb5
github.com/gliderlabs/logspout/router.(*RouteManager).Run.func1(0xc420024440, 0xc4200cbe40)
        /go/src/github.com/gliderlabs/logspout/router/routes.go:186 +0x37
created by github.com/gliderlabs/logspout/router.(*RouteManager).Run
        /go/src/github.com/gliderlabs/logspout/router/routes.go:185 +0xea
```

As you can see above, if the retry happens when the endpoint has recovered itself, it works fine. However, if the retry happens when the endpoint is still down, there's a panic. I will review the code when I have time, shouldn't be too tough to fix.

File: adapters/raw/raw.go [hunk 4]
```diff
@@ -7,14 +7,43 @@ import (
 	"log"
 	"net"
 	"os"
-	"reflect"
 	"text/template"
 
 	"github.com/gliderlabs/logspout/router"
+	"strconv"
```
Comment (gbolo) [inline adapters/raw/raw.go]: please run `goimports` on this file

File: adapters/raw/raw.go [hunk 5]
```diff
@@ -73,9 +104,90 @@ func (a *Adapter) Stream(logstream chan *router.Message) {
...
+
+func retryExp(fun func() error, tries uint) error {
+	try := uint(0)
+	for {
+		err := fun()
+		if err == nil {
+			return nil
+		}
+
+		try++
```
Comment (gbolo) [inline adapters/raw/raw.go]: we should log a message (possibly in debug) every time we retry, with the current value of `try`

File: adapters/raw/raw.go [hunk 6]
```diff
@@ -7,14 +7,43 @@ import (
...
 	"os"
-	"reflect"
 	"text/template"
 
 	"github.com/gliderlabs/logspout/router"
+	"strconv"
+	"time"
 )
 
+const defaultRetryCount = 10
```
Comment (gbolo) [inline adapters/raw/raw.go]: consider increasing this default, as with 10 it's only a few seconds of retries

File: adapters/raw/raw.go [hunk 7]
```diff
@@ -73,9 +104,90 @@ func (a *Adapter) Stream(logstream chan *router.Message) {
...
 		if err != nil {
 			log.Println("raw:", err)
-			if reflect.TypeOf(a.conn).String() != "*net.UDPConn" {
-				return
+			switch a.conn.(type) {
+			case *net.UDPConn:
+				continue
+			default:
+				if err = a.retry(buf.Bytes(), err); err != nil {
+					log.Panicf("raw retry err: %+v", err)
```
Comment (gbolo) [inline adapters/raw/raw.go]: clearly log that the max retries have been reached before you exit here

**Pull request closed without merging**